docs/src/lucet-runtime/killswitch.md

```diff
@@ -0,0 +1,346 @@
+# `KillSwitch`
+
+`KillSwitch` is a mechanism by which users of `lucet-runtime` can
+asynchronously request, very sternly, that an instance be disabled from
+running. If the instance is currently running, it will be stopped as soon as
+possible, and if the instance has not yet run, it will immediately exit with
+error when the `lucet` embedder attempts to run it. In some circumstances,
+however, a `KillSwitch` may successfully fire to no actual effect at any point
+in the program - one such example is a timeout in a hostcall that eventually
+faults; since timeouts cannot preempt hostcalls, the timeout may never be
+witnessed if the fault causes the host to never resume the Lucet instance.
+
+In this chapter we will describe both a typical usage of `KillSwitch` as a
+mechanism to enforce execution time limits, and the implementation complexities
+underlying `KillSwitch` correctness.
+
+`KillSwitch` are valid for termination only on the instance call after they are
+created, and until an instance is reset. When a call into a guest returns, the
+shared state by which `KillSwitch` signal is replaced, and an attempt to
+`terminate` will fail with an `Err`.
+
+## Example
+
+This example taken from `lucet_runtime_tests::timeout::timeout_in_guest`:
+```rust
+let module = mock_timeout_module();
+let region = TestRegion::create(1, &Limits::default()).expect("region can be created");
+let mut inst = region
+    .new_instance(module)
+    .expect("instance can be created");
+let kill_switch = inst.kill_switch();
+
+// Spawn a thread to terminate the instance after waiting for 100ms.
+let t = thread::Builder::new()
+    .name("killswitch".to_owned())
+    .spawn(move || {
+        thread::sleep(Duration::from_millis(100));
+        assert_eq!(kill_switch.terminate(), Ok(KillSuccess::Signalled));
+    })
+    .expect("can spawn a thread");
+
+// Begin running the instance, which will be terminated remotely by the KillSwitch.
+match inst.run("infinite_loop", &[]) {
+    Err(Error::RuntimeTerminated(TerminationDetails::Remote)) => {
+        // the result of a guest that was remotely terminated (a KillSwitch at work)
+    }
+    res => panic!("unexpected result: {:?}", res),
+}
+
+// wait for the KillSwitch-firing thread to complete
+t.join().unwrap();
+```
+
+## Implementation
+
+As this section discusses implementation details of `lucet-runtime`, it
+will refer to structures and enums that are unexported.  Since there are
+no `lucet-runtime` paths for these items, this section will occasionally
+use fully-qualified paths to the `lucet_runtime_internals` path where.
+they are originally defined.
```

Comment:
```suggestion
no `lucet-runtime` paths for these items, this section will occasionally
use fully-qualified paths to the `lucet_runtime_internals` path where
they are originally defined.
```

Small `.` nit.

Aside from that, my technical writer brain does not like `occasionally`. Maybe we can clarify which things we're using `lucet-runtime-internals` paths. (`KillState`, and `Domain` IIRC)

Reply:
I mostly reworked this paragraph and it now just explicitly says `Domain` and `KillState` are not public, and where to find them in `lucet_runtime_internals`.